### PR TITLE
Implement API

### PR DESCRIPTION
**Pull request by valentin-chaplean**
This change is [Reviewable](https://reviewable.io/reviews/chaplean/kanbanize-api-client-bundle/2)

### CODE REVIEW

File: Api/KanbanizeApi.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ * @method Route getPredecessorCards()  Get a list of predecessor cards            ( #/Predecessor_Cards/getPredecessorCards )
+ * @method Route putPredecessorCard()   Make a card a predecessor of a given card  ( #/Predecessor_Cards/addPredecessorCard )
+ * @method Route getSuccessorCards()    Get a list of successor cards              ( #/Successor_Cards/getSuccessorCards )
+ * @method Route putSuccessorCard()     Make a card a successor of a given card    ( #/Successor_Cards/addSuccessorCard )
+ *
+ * @package   Chaplean\Bundle\KanbanizeApiClientBundle\Api
+ * @author    Valentin - Chaplean <valentin@chaplean.coop>
+ * @copyright 2014 - 2018 Chaplean (http://www.chaplean.coop)
+ */
+class KanbanizeApi extends AbstractApi
```
Comment (mdevlamynck) [inline Api/KanbanizeApi.php]: Wouldn't specify v2 be clearer and prevent breaking changes issues the day a potential v3 comes around?

**Review by mdevlamynck — APPROVED** [review]

File: Api/KanbanizeApiV1.php [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+namespace Chaplean\Bundle\KanbanizeApiClientBundle\Api;
+
+use Chaplean\Bundle\ApiClientBundle\Api\AbstractApi;
+use Chaplean\Bundle\ApiClientBundle\Api\Parameter;
+use Chaplean\Bundle\ApiClientBundle\Api\Route;
+use GuzzleHttp\ClientInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class KanbanizeApiV1.
```
Comment (tomchaplean) [inline Api/KanbanizeApiV1.php:12]: Naming is not good. We need to have Api Suffix, so my proposal is KanbanizeLegacyApi and a comment to specify that it's V1

File: Api/KanbanizeApiV1.php [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * @method Route getProjectsAndBoards()   Get projects and boards     ( https://kanbanize.com/api/#get_projects_and_boards )
+ * @method Route postBoardStructure()     Get board structure         ( https://kanbanize.com/api/#get_board_structure )
+ * @method Route postFullBoardStructure() Get full board structure    ( https://kanbanize.com/api/#get_full_board_structure )
+ * @method Route postBoardSettings()      Get board settings          ( https://kanbanize.com/api/#get_board_settings )
+ * @method Route postBoardActivities()    Get board activities        ( https://kanbanize.com/api/#get_board_activities )
```
Comment (tomchaplean) [inline Api/KanbanizeApiV1.php:18]: Are you sure that post makes get ??

File: Api/KanbanizeApiV1.php [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+use GuzzleHttp\ClientInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class KanbanizeApiV1.
+ *
+ * @method Route getProjectsAndBoards()   Get projects and boards     ( https://kanbanize.com/api/#get_projects_and_boards )
+ * @method Route postBoardStructure()     Get board structure         ( https://kanbanize.com/api/#get_board_structure )
+ * @method Route postFullBoardStructure() Get full board structure    ( https://kanbanize.com/api/#get_full_board_structure )
+ * @method Route postBoardSettings()      Get board settings          ( https://kanbanize.com/api/#get_board_settings )
```
Comment (tomchaplean) [inline Api/KanbanizeApiV1.php:17]: Are you sure that post makes get ??

File: Api/KanbanizeApiV1.php [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+ * @method Route postBoardActivities()    Get board activities        ( https://kanbanize.com/api/#get_board_activities )
+ * @method Route postCreateTask()         Create a new task           ( https://kanbanize.com/api/#create_new_task )
+ * @method Route postTaskDetails()        Get details for one task    ( https://kanbanize.com/api/#get_task_details )
+ * @method Route postTasksDetails()       Get details for many task   ( https://kanbanize.com/api/#get_task_details )
+ * @method Route postAllTasks()           Get all tasks for one board ( https://kanbanize.com/api/#get_all_tasks )
+ * @method Route postLogTimeActivities()  Create a new task           ( https://kanbanize.com/api/#get_log_time_activities )
+ *
+ * @package   Chaplean\Bundle\KanbanizeApiClientBundle\Api
+ * @author    Valentin - Chaplean <valentin@chaplean.coop>
+ * @copyright 2014 - 2018 Chaplean (http://www.chaplean.coop)
```
Comment (tomchaplean) [inline Api/KanbanizeApiV1.php]: 2019 / https !

File: Api/KanbanizeApiV1.php [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * KanbanizeApiV1 constructor.
+     *
+     * @param ClientInterface          $client
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param string                   $urlPrefix
+     * @param string                   $apikey
+     */
+    public function __construct(ClientInterface $client, EventDispatcherInterface $eventDispatcher, string $urlPrefix, string $apikey)
```
Comment (tomchaplean) [inline Api/KanbanizeApiV1.php:45]: Return typehinting missing

File: Api/KanbanizeApiV2.php [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+namespace Chaplean\Bundle\KanbanizeApiClientBundle\Api;
+
+use Chaplean\Bundle\ApiClientBundle\Api\AbstractApi;
+use Chaplean\Bundle\ApiClientBundle\Api\Parameter;
+use Chaplean\Bundle\ApiClientBundle\Api\Route;
+use GuzzleHttp\ClientInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class KanbanizeApiV2.
```
Comment (tomchaplean) [inline Api/KanbanizeApiV2.php:12]: Name is not good too. My proposals are KanbanizeApi or KanbanizeCurrentApi or KanbanizeNewApi or KanbanizeV2Api if you really want to put the version in the name

File: Api/KanbanizeApiV2.php [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+ *
+ * @method Route getBoards()            Get a list of boards                       ( #/Boards/getBoards )
```
Comment (tomchaplean) [inline Api/KanbanizeApiV2.php]: We need urls, not paths, like in V1 API

File: Api/KanbanizeApiV2.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+ * @method Route getParentCards()       Get a list of parent cards                 ( #/Parent_Cards/getParentCards )
+ * @method Route putParentCard()        Make a card a parent of a given card       ( #/Parent_Cards/addParentCard )
+ * @method Route getPredecessorCards()  Get a list of predecessor cards            ( #/Predecessor_Cards/getPredecessorCards )
+ * @method Route putPredecessorCard()   Make a card a predecessor of a given card  ( #/Predecessor_Cards/addPredecessorCard )
+ * @method Route getSuccessorCards()    Get a list of successor cards              ( #/Successor_Cards/getSuccessorCards )
+ * @method Route putSuccessorCard()     Make a card a successor of a given card    ( #/Successor_Cards/addSuccessorCard )
+ *
+ * @package   Chaplean\Bundle\KanbanizeApiClientBundle\Api
+ * @author    Valentin - Chaplean <valentin@chaplean.coop>
+ * @copyright 2014 - 2018 Chaplean (http://www.chaplean.coop)
```
Comment (tomchaplean) [inline Api/KanbanizeApiV2.php]: 2019 / https !

File: Api/KanbanizeApiV2.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * KanbanizeApi constructor.
+     *
+     * @param ClientInterface          $client
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param string                   $urlPrefix
+     * @param string                   $apikey
+     */
+    public function __construct(ClientInterface $client, EventDispatcherInterface $eventDispatcher, string $urlPrefix, string $apikey)
```
Comment (tomchaplean) [inline Api/KanbanizeApiV2.php:44]: Return typehint missing

File: Api/KanbanizeApiV2.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            ])
+            ->urlPrefix($this->urlPrefix);
+
+        // Boards
+
+        $this->get('boards', '/boards')
+            ->queryParameters([
+                'board_ids'                       => Parameter::arrayList(Parameter::int())->optional(),
+                'project_ids'                     => Parameter::arrayList(Parameter::int())->optional(),
+                'is_archived'                     => Parameter::enum([0, 1])->optional(),
```
Comment (tomchaplean) [inline Api/KanbanizeApiV2.php:73]: Maybe it's time to create a boolean parameter, more explicit?

File: Resources/config/routing.yml [hunk 12]
```diff
@@ -1,3 +1,3 @@
 #kanbanize_client_index:
 #    path: /
-#    defaults: { _controller: ChapleanKanbanizeClientBundle:Default:index }
+#    defaults: { _controller: ChapleanKanbanizeApiClientBundle:Default:index }
```
Comment (tomchaplean) [inline Resources/config/routing.yml]: We can remove that, we don't need this file

Comment (tomchaplean) [inline Resources/config/routing.yml]: Config needs to be in config/ directory, not Resources anymore

File: Resources/config/services.yml [hunk 13]
```diff
@@ -1 +1,16 @@
-#services:
+services:
```
Comment (tomchaplean) [inline Resources/config/services.yml:1]: There is no such files in bundle : what is the purpose?

File: Tests/Api/KanbanizeApiV2Test.php [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+use GuzzleHttp\Psr7\Response;
+use Mockery\Adapter\Phpunit\MockeryTestCase;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class KanbanizeApiTest.
+ *
+ * @package   Chaplean\Bundle\KanbanizeApiClientBundle\Tests\Api
+ * @author    Valentin - Chaplean <valentin@chaplean.coop>
+ * @copyright 2014 - 2019 Chaplean (http://www.chaplean.coop)
```
Comment (tomchaplean) [inline Tests/Api/KanbanizeApiV2Test.php]: https

File: composer.json [hunk 15]
```diff
@@ -11,31 +11,26 @@
...
-        }
-    ],
     "require": {
         "php": ">=7.1",
         "symfony/config": "^3.0 || ^4.0",
         "symfony/console": "^3.0 || ^4.0",
         "symfony/dependency-injection": "^3.0 || ^4.0",
         "symfony/yaml": "^3.0 || ^4.0",
-        "chaplean/rest-client-bundle": "^4.0"
+        "chaplean/api-client-bundle": "v1.1"
```
Comment (tomchaplean) [inline composer.json]: Why prevent upgrade from minors versions? Should be "^1.1"

**Review by tomchaplean — CHANGES_REQUESTED** [review]

File: Api/KanbanizeApiV1.php (hunk 3 above)
Comment (valentin-chaplean) [reply Api/KanbanizeApiV1.php:18]: absolutely
`{
    "status": false,
    "error": "Unknown method. Please make sure you are sending an HTTP POST request."
}`

File: Api/KanbanizeApiV1.php (hunk 4 above)
Comment (valentin-chaplean) [reply Api/KanbanizeApiV1.php:17]: absolutely
`{
    "status": false,
    "error": "Unknown method. Please make sure you are sending an HTTP POST request."
}`

File: Api/KanbanizeApiV1.php (hunk 6 above)
Comment (valentin-chaplean) [reply Api/KanbanizeApiV1.php:45]: Return typehinting on the constructor isn't allowed

File: Resources/config/routing.yml (hunk 12 above)
Comment (valentin-chaplean) [reply Resources/config/routing.yml]: indeed :)

File: Resources/config/services.yml (hunk 13 above)
Comment (valentin-chaplean) [reply Resources/config/services.yml:1]: Services definition is necessary for use API classes.
Symfony doesn't ban the services in bundles: 
https://symfony.com/doc/current/bundles/extension.html
https://symfony.com/doc/current/bundles/best_practices.html#directory-structure

File: Resources/config/services.yml [hunk 16]
```diff
@@ -1 +1,16 @@
...
+    _defaults:
+        autowire: true
+        public: true
```
Comment (tomchaplean) [inline Resources/config/services.yml]: Services has to be private by default

**Review by tomchaplean — CHANGES_REQUESTED** [review]

**Review by tomchaplean — APPROVED** [review]

**Pull request merged**